packages/strapi/lib/core/bootstrap.js

```diff
@@ -287,31 +287,60 @@ module.exports = function(strapi) {
   strapi.config.port = _.get(strapi.config.currentEnvironment, 'server.port') || strapi.config.port;
   strapi.config.host = _.get(strapi.config.currentEnvironment, 'server.host') || strapi.config.host;
 
-  let hostname = strapi.config.host;
+  // Defines serverUrl value
+  let serverUrl = _.get(strapi.config.currentEnvironment, 'server.url', '');
+  serverUrl = _.trim(serverUrl, '/ ');
+  if (typeof serverUrl !== 'string') {
+    strapi.stopWithError(new Error('Invalid server url config. Make sure the url is a string.'));
+  }
+  if (serverUrl.startsWith('http')) {
+    try {
+      serverUrl = _.trim(new URL(strapi.config.currentEnvironment.server.url).toString(), '/');
+    } catch (e) {
+      strapi.stopWithError(
+        new Error('Invalid server url config. Make sure the url defined in server.js is valid.')
+      );
+    }
+  } else if (serverUrl !== '') {
+    serverUrl = `/${serverUrl}`;
+  }
+
+  // Defines adminUrl value
+  let adminUrl = _.get(strapi.config.currentEnvironment, 'server.admin.url', '/admin');
+  adminUrl = _.trim(adminUrl, '/ ');
+  if (typeof adminUrl !== 'string') {
+    throw new Error('Invalid admin url config. Make sure the url is a non-empty string.');
+  }
+  if (adminUrl.startsWith('http')) {
+    try {
+      adminUrl = _.trim(new URL(adminUrl).toString(), '/');
+    } catch (e) {
+      strapi.stopWithError(
+        e,
+        'Invalid admin url config. Make sure the url defined in server.js is valid.'
+      );
+    }
+  } else {
+    adminUrl = `${serverUrl}/${adminUrl}`;
+  }
+
+  // Defines adminPath value
+  let adminPath = adminUrl;
   if (
-    strapi.config.environment === 'development' &&
-    ['127.0.0.1', '0.0.0.0'].includes(strapi.config.host)
+    serverUrl.startsWith('http') &&
+    adminUrl.startsWith('http') &&
+    new URL(adminUrl).origin === new URL(serverUrl).origin
   ) {
-    hostname = 'localhost';
+    adminPath = adminUrl.replace(getCommonBeginning(serverUrl, adminUrl), '');
```

Comment:
Would it be possible to directly use here too ? `new URL(adminUrl).pathname;`

Reply:
Not really, to handle that case :
```
{
   url: "https://mywebsite.fr/strapi",
   url: "https://mywebsite.fr/strapi/admin"
}
```
In that case we want the admin path to just be `/admin`